ocaml/xapi/xapi_xenops.ml

```diff
@@ -2029,48 +2025,107 @@ let on_xapi_restart ~__context =
       ) (all_known_xenopsds ())
     |> List.flatten in
 
+  (* The list of VMs xenopsd knows about that (xapi knows about at all,
+     xapi has no idea about at all) *)
   let xenopsd_vms_in_xapi, xenopsd_vms_not_in_xapi =
     List.partition (fun ((id, _), _) ->
         try vm_of_id ~__context id |> ignore; true with _ -> false
       ) vms_in_xenopsds in
 
+  (* Of the VMs xapi knows about, partition that set into VMs xapi believes
+     should be running here, and those that it didn't *)
+  let xapi_thinks_are_here, xapi_thinks_are_not_here =
+    List.partition (fun ((id, _), _) ->
+        List.exists (fun (id', _) -> id=id') resident_vms_in_db)
+        xenopsd_vms_in_xapi in
+
+  (* Of those xapi thinks aren't here, are any running on another host? If
+     so, kill the VM here. If they aren't running on another host (to the
+     best of our knowledge), set the resident_on to be here. *)
+  let xapi_thinks_are_elsewhere, xapi_thinks_are_nowhere =
+    List.partition (fun ((id, _), _) ->
+      let vm_ref = vm_of_id ~__context id in
+      Db.is_valid_ref __context (Db.VM.get_resident_on ~__context ~self:vm_ref)
+    ) xapi_thinks_are_not_here in
+
+  (* This is the list of VMs xapi thought were running here, but actually
+     aren't *)
   let xapi_vms_not_in_xenopsd =
     List.filter (fun (id, _) ->
         not (List.exists (fun ((id', _), _) -> id' = id) vms_in_xenopsds)
       ) resident_vms_in_db in
 
-  (* Destroy any VMs running that aren't in Xapi's database *)
+  (* Log the state before we do anything *)
+  let maybe_log_em msg l =
+    if List.length l > 0 then begin
+      debug "%s" msg;
```

Comment:
I think this is a valid comment, though the log lines are easy to find as the thread id is part of the log line. I'll put a meaningful prefix on each line to make it more obvious though.